### PR TITLE
Add schedule feature and schedule command

### LINKED ISSUE #166: Add `schedule` command

### PR DESCRIPTION
**Pull request by geoffong11**
Format: schedule INDEX i/YYYY-MM-DD-HH-MM [a/SCHEDULENAME].

First, the `ScheduleCommandParser` object parses the command and creates a `ScheduleDescriptor` object (under ScheduleCommand as a static class) using Schedule, ScheduleName, and LastModifiedTime objects. Suppose ScheduleName is not specified, the meeting will be called `Meeting` Then it creates a ScheduleCommand that takes in the Index of Person to change and the ScheduleDescriptor. Through ScheduleCommand#execute(), replaces the original Person object (by Index), after editing the Schedule, ScheduleName, and LastModifiedDateTime object in the Person through a secondary constructor, that takes in Optional<Schedule> and Optional<ScheduleName>. Confirmation string is then printed.

`Optional` wrapper is used for Schedule and ScheduleName objects in Person object since upon initialization, they technically do not have a Schedule. I don't know if that is the best way forward, but I guess for ClearSchedule, it will be easier to implement (put in Optional.empty() as the variables). 

Edit (by @xavierpok) :
Closes #166

### CODE REVIEW

**Review by imkwokyong — COMMENTED** [review]
After tinkering with the code, I found some bugs:
1. Invalid datetime like `2020-00-01-01-30` and `2020-01-31-25-30` throws exceptions that are not handled implicitly
2. For the February month, is there a reason why `2021-02-31...` defaults to `2021 Feb 28 ...` in the personListPanel instead of saying invalid date time to the user?

image


Suggestions:
1. Check whether provided datetime is later than current datetime; if no, throw error message saying something like, "The provided datetime is over long ago' or something like that.

**Comment by xavierpok** [comment]
Closes #166

File: src/main/java/connexion/logic/commands/ScheduleCommand.java [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        return new ToStringBuilder(this)
+                .add("index", index)
+                .add("scheduleDescriptor", scheduleDescriptor)
+                .toString();
+    }
+
+    /**
+     * Stores the details of a scheduled meeting with the person.
+     */
+    public static class ScheduleDescriptor {
```
Comment (xavierpok) [inline src/main/java/connexion/logic/commands/ScheduleCommand.java:131]: Not a suggestion, but next time we talk about this, we can table somehow unifying the (with my PR) 3 classes that use some kind of static nested `Descriptor` under an interface. Not a pressing issue, though.

File: src/main/java/connexion/model/person/Schedule.java [hunk 2]
```diff
@@ -1,29 +1,90 @@
...
- * Class representing a schedule in Person. TO BE IMPLEMENTED
+ * Class representing a schedule in Person.
  */
-public abstract class Schedule implements PersonListDetailField<LocalDateTime> {
+public class Schedule implements PersonListDetailField<LocalDateTime> {
+
+    public static final String MESSAGE_CONSTRAINTS = "Schedule time should be in YYYY-MM-DD-HH-MM";
+
+    public static final DateTimeFormatter SCHEDULE_FORMATTER =
+            DateTimeFormatter.ofPattern("yyyy-MM-dd-HH-mm");
```
[the remark this answers is not in the document]
Comment (xavierpok) [reply src/main/java/connexion/model/person/Schedule.java:20]: Will probably throw some kind of `[DateTimeParseException](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeParseException.html)`

[the remark this answers is not in the document]
Comment (xavierpok) [reply src/main/java/connexion/model/person/Schedule.java]: implemented

**Comment by xavierpok** [comment]
> 1. Invalid datetime like `2020-00-01-01-30` and `2020-01-31-25-30` throws exceptions that are not handled implicitly
Now throws a ParseException but still needs a test case added for this 
Checked by `isValid` now 
> 2. For the February month, is there a reason why `2021-02-31...` defaults to `2021 Feb 28 ...` in the personListPanel instead of saying invalid date time to the user?
Handled by changing of `ResolverStyle `to STRICT

>3. Check whether provided ......

Will open issue collating suggested feature expansions after schedule PR merged

File: src/main/java/connexion/logic/commands/ScheduleCommand.java [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+                scheduleDescriptor.getLastModifiedDateTime();
+        // While semantically, it would make sense that this would always be changed,
+        // We do it like this for consistency with other fields
+        // And to move responsibility for updating this field to the parser,
+        // Like the other fields.
+        Optional<Schedule> updatedSchedule = Optional.ofNullable(scheduleDescriptor.getSchedule());
+        Optional<ScheduleName> updatedScheduleName = Optional.ofNullable(scheduleDescriptor.getScheduleName());
+        return new Person(name, phone, email, company, job, tags,
+                updatedSchedule, updatedScheduleName, updatedLastModifiedDateTime);
+    }
```
Comment (xavierpok) [inline src/main/java/connexion/logic/commands/ScheduleCommand.java:102]:
```suggestion
        Person toReturn =  new Person(name, phone, email, company, job, tags,
                updatedSchedule, updatedScheduleName, updatedLastModifiedDateTime);
        if (personToEditSchedule.getValue()) { // if the person was marked
            toReturn.mark();
        } else { // if the person was unmarked
            toReturn.unmark();
        }
        return toReturn;
    }
```

Comment (xavierpok) [reply src/main/java/connexion/logic/commands/ScheduleCommand.java:102]: Should add mark copying. Above code block should resolve.

Comment (xavierpok) [reply src/main/java/connexion/logic/commands/ScheduleCommand.java:102]: Tests currently (probably) don't check for mark copying. Should add as a test case also.

**Review by xavierpok — CHANGES_REQUESTED** [review]

**Review by xavierpok — COMMENTED** [review]
All outstanding issues raised in an earlier comment look resolved, LGTM, but merge conflicts with PR #172 need to be resolved. Suggest waiting for resolution before merging.

**Review by xavierpok — APPROVED** [review]
Forgot to approve, LGTM

**Review by imkwokyong — APPROVED** [review]
LGTM! Thanks for fixing the aforementioned bugs and the merge conflicts!

**Pull request merged**